test/real_content/test_swupd.bats

```diff
@@ -0,0 +1,211 @@
+# because we aren't going to test all swupd functionalities. The idea is
+# to test major swupd commands and standard use cases
+
+#TODO: Add test for crossing a format barrier
+
+# Optional parameters
+# $URL: Clearlinux content and version urls
+# $MAX_VERSIONS: The maximum number of versions we are going to be looking to test updates. Should be # at least 2 (current + update). Default is 15.
+
+local SWUPD_OPTS # Swupd options to use in tests
+#local FORMAT # Format of the last available version
+local VERSIONS # Array of last available versions. First item is the oldest
+local ROOT_DIR # Root directory of the installation
+
+# get_format: Echos the format of the version in $1
+# TODO: This is not the best way to handle that. we should have in swupd
+# codebase the information about the supported format of this swupd.
+get_format() {
+	v="$1"
+	curl -f ${URL}/${v}/format 2>/dev/null
+	return 0
+}
+
+# get_format_from_versions: Echos the first format available in a version from
+# a list of versions
+get_format_from_versions() {
+	for v in "$@"; do
+		format=$(get_format $v)
+		if [ -z "$format" ]; then # if version isn't published, discard
+			continue
+		else
+			echo $format
+			break
+		fi
+	done
+}
+
+test_setup() {
+	# Fill defaults in optional parameters
+	if [ -z "$URL" ]; then
+		URL=https://cdn.download.clearlinux.org/update
+	fi
+	if [ -z "$MAX_VERSIONS" ]; then
+		MAX_VERSIONS=15
+	fi
+
+	# Using github clr-bundles project to get official clear releases
+	#TODO: Discover available versions to work with custom mixers
+	#Depends on: https://github.com/clearlinux/mixer-tools/issues/467
+	version_list=$(git ls-remote --tags https://github.com/clearlinux/clr-bundles.git | grep -v {} | grep -v latest | sed "s/.*refs\/tags\///" | sort -gr | head -n $MAX_VERSIONS)
+	if [ -z "$version_list" ]; then
+		echo "Impossible to get versions list"
+		return 1
+
+	fi
+
+	# Get the last format
+	export FORMAT=$(get_format_from_versions $version_list)
+	version_list=$(echo "$version_list" | sort -g)
+	if [ -z "$FORMAT" ]; then
+		echo "Impossible to guess format"
+		return 1
+
+	fi
+
+	# Fill a list of last available versions (up to ${MAX_VERSIONS})
+	i=0
+	for v in $version_list; do
+		format=$(get_format $v)
+		# if version isn't published or is from a different version, discard
+		if [ -z "$format" ] || [ "$FORMAT" -ne "${format}" ]; then
+			continue
+		fi
+
+		VERSION[$i]=$v
+		i=$(($i+1))
+	done
+
+	# Not enough versions to run the test
+	if [ -z "${VERSION[0]}" ] || [ -z "${VERSION[1]}" ]; then
+		#TODO: This is going to break on first release in a new format
+		echo "We need at least 2 versions in format $FORMAT to continue with this test"
+		return 1
+
+	fi
+
+	sudo rm -rf "$TEST_NAME"
+	sudo mkdir -p "$TEST_NAME"
+	ROOT_DIR="${PWD}/${TEST_NAME}"
+
+	SWUPD_OPTS="-u ${URL} -p ${ROOT_DIR} -S ${ROOT_DIR}/var/lib/swupd/"
+}
+
+test_teardown() {
+	if [ "$DEBUG_TEST" != true ]; then
```

Comment:
I don't think this is very useful because it will only keep the state of the last test to run (update from first to last). For instance, if one of the 'incremental updates' tests fail, its test directory will be deleted by the 'update from first to last' test setup.

Reply:
I followed the same approach of the teardown in the testlib. So we have a problem in all tests with more than one test when DEBUG_TEST is used.